<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from IPython.core.display import display, HTML
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score


C:\Users\Games\AppData\Local\Temp\ipykernel_28432\3918809254.py:3: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [2]:
display(HTML("<style>.container { width:88% !important; }</style>"))

In [3]:
try:
    forsen = pd.read_csv('/datasets/insurance.csv')
except:
    forsen = pd.read_csv('C:/Users/Games/Downloads/insurance.csv')

In [4]:
forsen.info(), forsen.shape, display(forsen.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

(None, (5000, 5), None)

In [5]:
print('Дубликатов в датасете:', forsen.duplicated().sum())

Дубликатов в датасете: 153


In [6]:
duplicateRows = forsen[forsen.duplicated()]
display(duplicateRows)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281     1     39.0   48100.0            1                  0
488     1     24.0   32900.0            1                  0
513     0     31.0   37400.0            2                  0
718     1     22.0   32600.0            1                  0
785     0     20.0   35800.0            0                  0
...   ...      ...       ...          ...                ...
4793    1     24.0   37800.0            0                  0
4902    1     35.0   38700.0            1                  0
4935    1     19.0   32700.0            0                  0
4945    1     21.0   45800.0            0                  0
4965    0     22.0   40100.0            1                  0

[153 rows x 5 columns]

In [7]:
forsen = forsen.drop_duplicates()

In [8]:
forsen.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                4847 non-null   int64  
 1   Возраст            4847 non-null   float64
 2   Зарплата           4847 non-null   float64
 3   Члены семьи        4847 non-null   int64  
 4   Страховые выплаты  4847 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 227.2 KB


In [9]:
X = forsen.drop(columns='Страховые выплаты', axis=1)
y = forsen['Страховые выплаты']

X.shape, y.shape

((4847, 4), (4847,))

Открыли файл, посмотрели, что в датасете 5к строк, 5 столбцов. Пропусков нет, было 153 дубликата, я их убрал. Не должно сказаться на дальнейших результатах работы. Признаки определили выше, целевой в том числе.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ: b (не изменится)**

**Обоснование: (ниже)**

Начнём с формулы обучения и заменим в ней X на XZ (умножение на обратимую матрицу) и далее раскроем скобки и приведем формулу к такому виду, чтобы можно было показать отношение между $w$ и $w_{new}$.

$$
a_{new} = XZw_{new}
$$

Представим новую матрицу признаков $X1$ как произведение старой $X$ на матрицу $Z$(обратимую):

$$
X1 = X * Z
$$

Подставим новое значение X1 в формулу $w_{new}$:

$$
w_{new} = ((XZ)^T XZ)^{-1} (XZ)^T y
$$

Раскроем первое произведение $(XZ)^T$:

$$
w_{new} = (Z^T X^T XZ)^{-1} Z^T X^T y 
$$

Перегруппируем множители в скобках:

$$
w_{new} = (Z^T (X^T X) Z)^{-1} Z^T X^T y
$$

Так как $Z$ по условию обратимая, то произведение $(Z^T)^{-1} Z^T = E$ (единичная матрица):

$$
w_{new} = Z^{-1} (X^T X)^{-1} E X^T y = Z^{-1} (X^T X)^{-1} X^T y 
$$

Можно заметить, что справа у нас оказалась формула для $w$:

$$
w_{new} = Z^{-1} (X^T X)^{-1} X^T y = Z^{-1} w
$$

Подставим новое значение весов 𝑤1 в формулу для предсказаний линейной регрессии:

$$
a_{new} = XZZ^{-1} w
$$

Так как $Z$ по условию обратимая, то произведение $ZZ^{-1}$ снова равно $E$:

$$
a_{new} = XZZ^{-1} w = XEw = Xw = a
$$

Предсказания $a_{new}$ для матрицы признаков, умноженных на обратимую матрицу $Z$ равны предсказаниям $a$. Качество лин. регрессии при умножении признаков на обратимую матрицу не изменится. Что касается связи для $w$ и $w_{new}$, то $w_{new} = Z^{-1} w$. 

## Алгоритм преобразования

**Алгоритм**

Для шифрования данных будем умножать матрицу признаков на обратимую матрицу, которая будет генерироваться случайным образом.

Этапы алгоритма:

* Составление матрицы;
* проверка матрицы на обратимость;
* получение матрицы преобразованных признаков;
* применение алгоритма на преобразованных признаках и сравнение полученного R2 score с R2 score на оригинальных данных;


**Обоснование**

При умножении признаков на обратимую матрицу качество линейной регрессии не изменится, то есть данный алгоритм позволяет защитить данные, а потом восстановить их, зная обратную матрицу P, если это необходимо

## Проверка алгоритма

In [10]:
def data_encrypted(X, y):
    
    Z = np.random.rand(X.shape[1], X.shape[1])
    while np.linalg.det(Z) == 0:
        Z = np.random.rand(X.shape[1], X.shape[1])
    X_encrypted = X@Z
    
    model = LinearRegression()
    model.fit(X_encrypted, y)
    
    y_pred_encrypted = model.predict(X_encrypted)
    r2_encrypted = r2_score(y, y_pred_encrypted)
    
    return X_encrypted, r2_encrypted


X_encrypted, r2_encrypted = data_encrypted(X, y)


model_original = LinearRegression()
model_original.fit(X, y)

y_pred_original = model_original.predict(X)
r2_original = r2_score(y, y_pred_original)

print("Original data R2 score:", r2_original)
print("Encrypted data R2 score:", r2_encrypted)

Original data R2 score: 0.4302010044852067
Encrypted data R2 score: 0.4302010044852015


**Вывод:**

**Метрики одинаковы. Качество линейной регрессии из sklearn не отличается до и после преобразования.**